### PR TITLE
JSON unescape of unicode chars to utf-8 was incorrect

### PR DESCRIPTION
**Pull request by otaviobp**
The implementation of sol_json_token_get_unescaped_string was converting
unicode number to utf-8 in an incorrect way. The conversion is a bit
more complex than what was implemented.

This patch fixes this issue and also adds a test to check unicode
conversions

Signed-off-by: Otavio Pontes <otavio.pontes@intel.com>

### CODE REVIEW

File: src/lib/parsers/sol-json.c [hunk 1]
```diff
@@ -690,6 +690,24 @@ sol_json_serialize_boolean(struct sol_buffer *buffer, bool val)
...
+static int
+_hex_decode(const char *p, uint8_t *val)
+{
+    int16_t n;
+    char hexdigits[3], *endptr;
+
+    hexdigits[2] = '\0';
+    memcpy(hexdigits, p, 2);
+    errno = 0;
+    n = strtoul(hexdigits, &endptr, 16);
```
Comment (lpereira) [inline src/lib/parsers/sol-json.c]: `strtoul()` returns an `unsigned long`, and the return value is being truncated to an `int16_t`.

Comment (lpereira) [inline src/lib/parsers/sol-json.c]: Maybe this is faster/clearer?

      bool
      parse_hex_digit(uint8_t digit, uint8_t *nibble)
      {
        if (digit >= '0' && digit <= '9') {
          *nibble = digit - '0';
          return true;
        }
        if (digit >= 'a' && digit <= 'f') {
          *nibble = digit - 'a';
          return true;
        }
        if (digit >= 'A' && digit <= 'F') {
          *nibble = digit - 'A';
          return true;
        }

        return false;
      }
   
      int
      parse_hex(char *in, uint8_t *out)
      {
        uint8_t nibble;

        if (!parse_hex_digit(in[0], &nibble))
          return -EINVAL;

        *out = nibble;

        if (!parse_hex_digit(in[1], &nibble))
          return -EINVAL;

        *out = (*out << 4) | nibble;

        return 0;
      }

Comment (otaviobp) [inline src/lib/parsers/sol-json.c]: not a bad idea

Comment (otaviobp) [inline src/lib/parsers/sol-json.c]: And strtoul returns negative numbers casted as unsined int. I am getting it as a signed int to check for negative values. But anyway I prefer your other solution

Comment (lpereira) [inline src/lib/parsers/sol-json.c]: The thing is: they're being truncated to a 16-bit int from a 64-bit unsigned int. That's not going to work well. :)
